File: projects/dejavu.cs/Dejavu.Language/ParseError.cs
namespace Dejavu.Language;

/// <summary>Span-aware parse failure.</summary>
public sealed class ParseError : Exception
{
    public string File { get; }
    public int Start { get; }
    public int Length { get; }
    public string Label { get; }

    public ParseError(
        string message,
        int start,
        int length = 1,
        string? file = null,
        string? label = null)
        : base(message)
    {
        File = file ?? "template.dejavu";
        Start = start;
        Length = Math.Max(1, length);
        Label = label ?? "here";
    }
}

File: projects/dejavu.kt/dejavu-language/src/main/kotlin/dejavu/language/ParseError.kt
package dejavu.language

/** Span-aware parse failure. */
class ParseError(
    message: String,
    val file: String = "template.dejavu",
    val start: Int,
    length: Int = 1,
    val label: String = "here",
) : RuntimeException(message) {
    val length: Int = maxOf(1, length)
}
